File: src/java/magma/compile/rule/TypeRule.java
package magma.compile.rule;

import jvm.collect.list.Lists;
import magma.compile.CompileError;
import magma.compile.Node;
import magma.compile.context.NodeContext;
import magma.compile.context.StringContext;
import magma.result.Err;
import magma.result.Result;

public record TypeRule(String type, Rule rule) implements Rule {
    @Override
    public Result<Node, CompileError> parse(String input) {
        return rule.parse(input)
                .mapValue(node -> node.retype(type))
                .mapErr(error -> new CompileError("Failed to assign type", new StringContext(input), Lists.of(error)));
    }

    @Override
    public Result<String, CompileError> generate(Node node) {
        if (node.is(type)) {
            return rule.generate(node);
        }

        return new Err<>(new CompileError("Node was not of type '" + type + "'", new NodeContext(node)));
    }
}

File: src/clang/magma/compile/rule/tree/TypeRule.c
#include "TypeRule.h"
auto __lambda0__(){return node;
}
auto __lambda1__(){return CompileError("Failed to parse of type '" + type + "'", StringContext(input), Lists.of(error));
}
magma.result.Result<magma.compile.Node, magma.compile.CompileError> parse(String input){return rule.parse(input).mapValue(__lambda0__.retype(type)).mapErr(__lambda1__);
}
magma.result.Result<String, magma.compile.CompileError> generate(magma.compile.Node node){if (node.is(type)) {
            return rule.generate(node).mapErr(err -> new CompileError("Failed to generate of type '" + type + "'", new NodeContext(node), Lists.of(err)));
        }return (CompileError("Node was not of type '" + type + "'", NodeContext(node)));
}
